staging/src/k8s.io/csi-translation-lib/plugins/in_tree_volume.go

```diff
@@ -59,3 +63,89 @@ type InTreePlugin interface {
 	// RepairVolumeHandle generates a correct volume handle based on node ID information.
 	RepairVolumeHandle(volumeHandle, nodeID string) (string, error)
 }
+
+// replaceTopology overwrites an existing topology by a new one.
+func replaceTopology(pv *v1.PersistentVolume, oldKey string, newTopology v1.NodeSelectorRequirement) error {
+	for i := range pv.Spec.NodeAffinity.Required.NodeSelectorTerms {
+		for j, r := range pv.Spec.NodeAffinity.Required.NodeSelectorTerms[i].MatchExpressions {
+			if r.Key == oldKey {
+				pv.Spec.NodeAffinity.Required.NodeSelectorTerms[i].MatchExpressions[j] = newTopology
+				return nil
+			}
+		}
+	}
+	return fmt.Errorf("could not find topology key %q", oldKey)
+}
+
+// getTopology returns the current topology zones with the given key.
+func getTopology(pv *v1.PersistentVolume, key string) []string {
+	if pv.Spec.NodeAffinity == nil ||
+		pv.Spec.NodeAffinity.Required == nil ||
+		len(pv.Spec.NodeAffinity.Required.NodeSelectorTerms) < 1 {
+		return nil
+	}
+	for i := range pv.Spec.NodeAffinity.Required.NodeSelectorTerms {
+		for _, r := range pv.Spec.NodeAffinity.Required.NodeSelectorTerms[i].MatchExpressions {
+			if r.Key == key {
+				return r.Values
+			}
+		}
+	}
+	return nil
+}
+
+// addTopology appends the topology to the given PV.
+func addTopology(pv *v1.PersistentVolume, newTopology v1.NodeSelectorRequirement) error {
+	// Make sure the necessary fields exist
+	if pv.Spec.NodeAffinity == nil {
```

Comment:
We can probably simplify this and assume in the "add" case, NodeAffinity is not set.

Reply:
Done